File: day4/day4.c
#include <assert.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>

int rule2(const char *str)
{
	int p = 0;
	for (const char *s = str; *s; s++)
	{
		if (*s == p)
		{
			return 1;
		}
		p = *s;
	}
	return 0;
}

int rule3(const char *str)
{
	int p = 0;
	for (const char *s = str; *s; s++)
	{
		if (*s < p)
		{
			return 0;
		}
		p = *s;
	}
	return 1;
}

int rule4(const char *str)
{
	int t = INT_MAX;
	int p = 0;
	int c = 0;
	for (const char *s = str; *s; s++)
	{
		if (*s == p)
		{
			c++;
		}
		else if (c > 0 && c < t)
		{
			t = c;
			c = 0;
		}
		else
		{
			c = 0;
		}
		p = *s;
	}
	if (c > 0 && c < t)
	{
		t = c;
	}
	return t == 1;
}

int main(int argc, char *argv[])
{
	assert(rule2("111111") && rule3("111111"));
	assert(rule2("223450") && !rule3("223450"));
	assert(!rule2("123789") && rule3("123789"));
	assert(rule4("112233"));
	assert(!rule4("123444"));
	assert(rule4("111122"));

	char buf[10];
	int p1, p2;
	p1 = p2 = 0;
	for (int i = 109165; i < 576723+1; i++)
	{
		snprintf(buf, sizeof(buf), "%d", i);
		int r3 = rule3(buf);
		if (rule2(buf) && r3)
		{
			p1++;
		}
		if (r3 && rule4(buf))
		{
			p2++;
		}
	}
	printf("part1: %d\npart2: %d\n", p1, p2);
	return 0;
}

File: day4/day4.py
#!/usr/bin/env python3

def rule2(s):
    p = None
    for a in s:
        if a == p:
            return True
        p = a
    return False

def rule3(s):
    p = "0"
    for a in s:
        if a < p:
            return False
        p = a
    return True

def rule4(s):
    t = len(s)
    p = None
    c = 0
    for a in s:
        if a == p:
            c += 1
        else:
            if c > 0 and c < t:
                t = c
            c = 0
        p = a

    if c > 0 and c < t:
        t = c

    return t == 1

x = 111111
assert rule2(str(x)) and rule3(str(x))
x = 223450
assert rule2(str(x)) and not rule3(str(x))
x = 123789
assert not rule2(str(x)) and rule3(str(x))
assert rule4("112233")
assert not rule4("123444")
assert rule4("11122")

p1 = 0
p2 = 0
for x in range(109165,576723+1):
    s = str(x)
    r3 = rule3(s)
    if rule2(s) and r3:
        p1 += 1
    if r3 and rule4(s):
        p2 += 1

print("part1", p1)
print("part2", p2)
